src/main/host/descriptor/mod.rs

```diff
@@ -436,9 +430,98 @@ impl std::fmt::Debug for PosixFileRefMut<'_> {
     }
 }
 
+/// Represents a POSIX file description, or a Linux `struct file`. An `OpenFile` wraps a reference
+/// to a `GenericFile`. Once there are no more `OpenFile` objects for a given `GenericFile`, the
+/// `GenericFile` will be closed. Typically this means that holding an `OpenFile` will ensure that
+/// the file remains open (the file's status will not become `FileStatus::CLOSED`), but the
+/// underlying file may close itself in extenuating circumstances (for example if the file has an
+/// internal error).
+///
+/// **Safety:** If an `OpenFile` for a specific `GenericFile` already exists, it is an error to
```

Comment:
> It'd be useful to document what structure or code convention we're using to (try to) ensure this.

I'm not sure how to enforce this. I think in almost all cases we should only create a new `OpenFile` object for a newly created file object, so I added a comment about this. But as for compile-time enforcement, I don't have any ideas for nice ways to do this.

> Maybe it'd be worth some runtime check like having `GenericFile` keep track of whether it already has an `OpenFile`? (Maybe only in debug builds...?)

I added a runtime check for this, and it will log a warning.